### PR TITLE
ignore status in shipment updates

### LINKED ISSUE #650: OS/IS: Update mutation shouldn't set status
Inbound & Outbound shipments aren't updating because they are trying to set the status, and only prescribed values are allowed for the status field, so the update fails.

The update mutation now allows partials updates, we can send only modified fields

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #650 

I now have a problem with focus - when you start typing into a field, pause, and the page rerenders the input, you lose focus and have to click in the text area/field again which is super annoying. have left that little issue for now.

### CODE REVIEW

File: packages/common/src/utils/arrays/utils.ts [hunk 1]
```diff
@@ -26,7 +26,7 @@ export const groupBy = <T, K extends keyof T & string>(
   key: K
 ): Record<string, T[]> => {
   const lookup: Record<string, T[]> = {} as Record<string, T[]>;
-  things.forEach(thing => {
+  things?.forEach(thing => {
```
Comment (mark-prins) [inline packages/common/src/utils/arrays/utils.ts:29]: offroad fix - was throwing an error if there were no lines

File: packages/invoices/src/OutboundShipment/DetailView/api.ts [hunk 2]
```diff
@@ -97,19 +98,30 @@ export const onRead =
     };
   };
 
+const getPatchStatus = (patch: Partial<OutboundShipment>) => {
```
Comment (mark-prins) [inline packages/invoices/src/OutboundShipment/DetailView/api.ts:101]: yeah. but it's temporary, right?

File: packages/invoices/src/OutboundShipment/DetailView/tabs/GeneralTab.tsx [hunk 3]
```diff
@@ -96,12 +96,12 @@ export const GeneralTabComponent: FC<
   const toggleGrouped = () => {
     setIsGroupedByItem({
       ...isGroupedByItem,
-      outboundShipment: !isGroupedByItem.outboundShipment,
+      outboundShipment: !isGroupedByItem?.outboundShipment,
```
Comment (mark-prins) [inline packages/invoices/src/OutboundShipment/DetailView/tabs/GeneralTab.tsx:99]: and the other offroad. oops!

File: packages/invoices/src/InboundShipment/DetailView/api.ts [hunk 4]
```diff
@@ -325,10 +325,7 @@ export const useInboundFields = <KeyOfInvoice extends keyof Invoice>(
           id,
         ]);
 
-        queryClient.setQueryData(['invoice', id], {
-          ...previousInbound,
-          ...patch,
-        });
+        queryClient.setQueryData(['invoice', id], patch);
```
Comment (joshxg) [inline packages/invoices/src/InboundShipment/DetailView/api.ts:328]: This is causing the loss of focus - The query data becomes just a partial inbound like `{ id, comment}`, which the whole page reacts to - there's no lines, so the table goes away, there's no other party so the other party input goes away then the status isn't an editable status, so the whole form is disabled causing the loss of focus. I've put it back to what is was previously and it's working fine and updating - but adding items isn't as that uses the `upsert` mutation which includes updating the shipment including the status..!

File: packages/invoices/src/OutboundShipment/DetailView/api.ts (hunk 2 above)
Comment (joshxg) [reply packages/invoices/src/OutboundShipment/DetailView/api.ts:101]: Yup! 😁

**Review by joshxg — APPROVED** [review]
I added this commit 535c53c

which is working well for me against the demo server api. Though also temporary as we can't change status with it!

File: packages/common/src/utils/arrays/utils.ts (hunk 1 above)
Comment (mark-prins) [reply packages/common/src/utils/arrays/utils.ts:29]: lucky me 🤷  Am using the demo server API and one of the invoices ( inbound, I think ) which has no lines was giving an error. just a console error, so no harm done.

File: packages/invoices/src/InboundShipment/DetailView/api.ts (hunk 4 above)
Comment (mark-prins) [reply packages/invoices/src/InboundShipment/DetailView/api.ts]: oh! 🤦  thanks

**Pull request merged**